### PR TITLE
Added overload for 'CollectionAssert' with IEqualityComparer.

### PR DESCRIPTION
**Pull request by dafenix**
Useful for comparisons of collections with slightly different objects. I hope you find it useful.

with kind regards
Daniel

### CODE REVIEW

**Comment by CharliePoole** [comment]
I'd prefer not to add this. We have always treated `CollectionAssert` as a stable, legacy API, created to match the Microsoft class of the same name for those who like the syntax. Everything one can do with `CollectionAssert` and more can be done using `Assert.That`.

That said, I don't ,now how the Microsoft `CollectionAssert` has evolved since I created our class. I'd be open to a review of the classes with some overall decisions about where it should go before making a final call on this PR. In particular, if we are adding such an overload to `AreEquivalent` users will expect it on other members as well.

Marking this as an idea till we have a concensus on overall direction.

**Comment by CharliePoole** [comment]
I made this a discussion item as well and assigned myself to facilitate the discussion. I'll remove myself the moment we make it a backlog item. All this is a little hard to follow on a PR, of course, which is why we generally like to have issues before we have code. However, let's give it a try.

**Comment by jnm2** [comment]
@CharliePoole This gives us an immediate fix for https://github.com/nunit/nunit/issues/2494 though. We just use a HashSet and rely on the hashes generated by the IEqualityComparer that is passed in. Then we tell people with a performance problem to provide their own IEqualityComparer instance and voila, we're as fast as the BCL itself.

**Comment by CharliePoole** [comment]
@jnm2 This has nothing to do with #2494, although a similar change could be made for that issue. I'd raise the same objection, however.

There is already a workaround for people who want to do this, which is to use `Assert.That`. I think we need to decide what our policy toward `CollectionAssert` really is before embellishing it.

**Comment by ChrisMaddock** [comment]
We need to be careful with that - we made changes a couple of releases ago so that CollectionAssert already behaves _differently_ to Assert-Is-Equal.

https://github.com/nunit/nunit/issues/2097#issuecomment-292729733

**Comment by CharliePoole** [comment]
@ChrisMaddock But issue #2494, brought up here by @jnm2, doesn't have to do with equality. It relates to the `UniqueItemsConstraint`, which works the same way whether you invoke it from `Assert.That` or from `CollectionAssert`.

**Comment by CharliePoole** [comment]
@nunit/framework-team Are you happy discussing the question "Should we add any enhancements to CollectionAssert?" here or would you prefer a separate issue?

**Comment by jnm2** [comment]
> This has nothing to do with #2494,

> doesn't have to do with equality

#2494 has to do with IEqualityComparer because without a hash code generator we have an inherent time complexity performance problem. This issue proposes to give us a hash code generator which in turn gives an obvious fix for #2494. Plus, uniqueness is concerned with equality.

**Comment by jnm2** [comment]
We don't have to affect `UniqueItemsConstraint` to use the API in this issue to fix #2494.

**Comment by CharliePoole** [comment]
@jnm2 Sorry, I'm not seeing what you mean. How can adding an overload to `CollectionConstraint.AreEquvalent` fix performance in a test that uses `CollectionConstraint.AllItemsAreUnique`? If you are saying a similar change could fix the latter, I agree and already said so - although I don't want to do it. I would rather fix the initial problem rather than giving users a tool to create a workaround, especially since it's trivial to do so.

**Comment by CharliePoole** [comment]
@jnm2 This is a seriously odd way to discuss #2694!

However... If we want `CollectionConstraint.AllItemsAreUnique` to work differently, we have to either change it or change the underlying UniqueItemsConstraint. 

Please lets stop discussing #2694 here. I recognize that I answered your reference to it and contributed to everyone else's confusion so I'm going to stop.

**Comment by jnm2** [comment]
@CharliePoole Oh! I erroneously thought he was talking about adding an overload to `AllItemsAreUnique`. You are quite right and I'm only causing confusion. And, I agree with you that we should direct people away from CollectionAssert.

**Comment by jnm2** [comment]
> This is a seriously odd way to discuss #2694!

Lol! It certainly is! I'm so bad at this.

**Comment by rprouse** [comment]
It has been our position for quite awhile now that we are not going to be enhancing the legacy asserts. The legacy asserts suffer from the fact that we need to add overloads or new methods for different usages like this whereas the constraint syntax can be combined. This code is functionally equivalent to `Is.EquivalentTo().Using()`, so it doesn't add new behavior, just another way to do the same thing.

@CharliePoole has raised that we might want to continue to evolve to match the MSTest API. This does make it easier to migrate, but I don't believe that NUnit should just be MSTest plus 😄 

Thanks for your contribution, but I am going to close this.
